cni/pkg/nodeagent/cni-watcher_test.go

```diff
@@ -216,3 +217,26 @@ func TestCNIPluginServerPrefersCNIProvidedPodIP(t *testing.T) {
 	// Assert expected calls actually made
 	fs.AssertExpectations(t)
 }
+
+type retryableHandler struct {
+	counter chan struct{}
+}
+
+func (r retryableHandler) GetPodIfAmbient(podName, podNamespace string) (*corev1.Pod, error) {
+	r.counter <- struct{}{}
+	return nil, nil
+}
+func (r retryableHandler) GetAmbientPods() []*corev1.Pod { return nil }
+func (r retryableHandler) Start()                        {}
+
+func TestReconcileCNIAddEventGetPodIfAmbientShouldRetryable(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "retry_ns"}}
+	client := kube.NewFakeClient(ns)
+	retryableHandler := retryableHandler{counter: make(chan struct{}, 20)}
+	meshDataplane := &meshDataplane{kubeClient: client.Kube()}
+	pluginServer := startCniPluginServer(ctx, "/tmp/test.sock", retryableHandler, meshDataplane)
+	_ = pluginServer.ReconcileCNIAddEvent(ctx, CNIPluginAddEvent{})
+	require.Greater(t, len(retryableHandler.counter), 1, "GetPodIfAmbient should not be executed only once due to the retry policy.")
```

Comment:
I like/prefer the `testify` assert libs too, but this `require` can be expressed with a plain `assert` from `"istio.io/istio/pkg/test/util/assert"` which is already imported, so we don't need to add a new dep.

Also, this would retry the max amount of times, correct? This is synchronous so we will always retry 10x, so I don't think anything is gained by checking "at least one"

Reply:
>Also, this would retry the max amount of times, correct? This is synchronous so we will always retry 10x, so I don't think anything is gained by checking "at least one"

Since the loop terminates with `(ambientPod == nil) && (retries <= maxStaleRetries)`. So I don't think there will be a case where it always executes 10 times. Also I think the elapsed time here should be around [0 ~ 10x10ms]. 0 is the first time a valid ambientPod is queried, 10x10ms is the retries until all the chances are consumed.

By "synchronization" do you mean the "counter channel" in "retryableHandler"? If so, since it's a 20 size buffer channel here, I don't think we need to worry about that, there won't be a synchronization block problem.

@bleggett 
If you want, I can also add cases to the unit test for the two scenarios _error occurred_ and _successful fetch_, maybe that might be better?

Thank you for your time.

Reply:
Sorry, it dawned on me that you were probably talking about retrying the test 10 times, which is true in unit tests.